x-pack/plugins/security/server/lib/authentication/providers/oidc.ts

```diff
@@ -0,0 +1,529 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import Boom from 'boom';
+import { Legacy } from 'kibana';
+import { canRedirectRequest } from '../../can_redirect_request';
+import { getErrorStatusCode } from '../../errors';
+import { AuthenticationResult } from '../authentication_result';
+import { DeauthenticationResult } from '../deauthentication_result';
+import { AuthenticationProviderOptions, BaseAuthenticationProvider } from './base';
+
+/**
+ * The state supported by the provider (for the OpenID Connect handshake or established session).
+ */
+interface ProviderState {
+  /**
+   * Unique identifier of the OpenID Connect request initiated the handshake used to mitigate
+   * replay attacks.
+   */
+  nonce?: string;
+
+  /**
+   * Unique identifier of the OpenID Connect request initiated the handshake used to mitigate
+   * CSRF.
+   */
+  state?: string;
+
+  /**
+   * URL to redirect user to after successful OpenID Connect handshake.
+   */
+  nextURL?: string;
+
+  /**
+   * Access token issued as the result of successful OpenID Connect handshake and that should be provided with
+   * every request to Elasticsearch on behalf of the authenticated user. This token will eventually expire.
+   */
+  accessToken?: string;
+
+  /**
+   * Once access token expires the refresh token is used to get a new pair of access/refresh tokens
+   * without any user involvement. If not used this token will eventually expire as well.
+   */
+  refreshToken?: string;
+}
+
+/**
+ * Defines the shape of an incoming OpenID Connect Request
+ */
+type OidcIncomingRequest = Legacy.Request & {
+  payload: { iss?: string };
+  query: { iss?: string; code?: string; state?: string };
+};
+
+/**
+ * Defines the parameters that can be passed to ES when calling the /oidc/prepare endpoint
+ */
+interface OidcPrepareParams {
+  nonce?: string;
+  state?: string;
+  iss?: string;
+  code?: string;
+  realm?: string;
+}
+
+/**
+ * Defines the realm specific properties
+ */
+type OpenIdConnectProviderOptions = AuthenticationProviderOptions & {
+  realm: string;
+};
+
+function isOidcIncomingRequest(request: Legacy.Request): request is OidcIncomingRequest {
```

Comment:
Thanks for your comment, it made me realize that I'm not handling error responses here at all. I'll fix that. 
I added a comment in the method doc to describe when we can get each query that we handle as an OIDC related one

```
/**
 * Checks if the Request object represents an HTTP request regarding authentication with OpenID
 * Connect. This can be
 * - An HTTP GET request with a query parameter named `iss` as part of a 3rd party initiated authentication
 * - An HTTP POST request with a parameter named `iss` as part of a 3rd party initiated authentication
 * - An HTTP GET request with a query parameter named `code` as the response to a successful authentication from
 *   an OpenID Connect Provider
 * - An HTTP GET request with a query parameter named `error` as the response to a failed authentication from
 *   an OpenID Connect Provider
 * @param request Request instance.
 */
```

Regarding your concern. These are requests hitting the /api/security/v1/oidc endpoint and as such if one comes with a `code` query parameter, I expect it to be as part of an OIDC authentication request. If this is not the case, Elasticsearch will simply fail to exchange the code for an ID Token from the OpenID Connect Provider and will return an error.